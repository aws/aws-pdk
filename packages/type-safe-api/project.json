{
  "name": "@aws/type-safe-api",
  "root": "packages/type-safe-api",
  "targets": {
    "build": {
      "inputs": [
        "default",
        "^default",
        "!{projectRoot}/.jsii",
        "!{projectRoot}/lib/**/*",
        "!{projectRoot}/dist/**/*",
        "!{projectRoot}/coverage/**/*",
        "!{projectRoot}/test-reports/**/*",
        "!{projectRoot}/docs/api/**/*"
      ],
      "outputs": [
        "{projectRoot}/.jsii",
        "{projectRoot}/lib",
        "{projectRoot}/dist",
        "{projectRoot}/coverage",
        "{projectRoot}/test-reports",
        "{projectRoot}/docs/api"
      ],
      "dependsOn": [
        "^build"
      ],
      "executor": "nx:run-commands",
      "options": {
        "command": "pnpm exec projen build",
        "cwd": "packages/type-safe-api"
      }
    },
    "default": {
      "executor": "nx:run-commands",
      "options": {
        "command": "pnpm exec projen default",
        "cwd": "packages/type-safe-api"
      }
    },
    "pre-compile": {
      "executor": "nx:run-commands",
      "options": {
        "command": "pnpm exec projen pre-compile",
        "cwd": "packages/type-safe-api"
      }
    },
    "compile": {
      "executor": "nx:run-commands",
      "options": {
        "command": "pnpm exec projen compile",
        "cwd": "packages/type-safe-api"
      }
    },
    "post-compile": {
      "executor": "nx:run-commands",
      "options": {
        "command": "pnpm exec projen post-compile",
        "cwd": "packages/type-safe-api"
      }
    },
    "test": {
      "executor": "nx:run-commands",
      "options": {
        "command": "pnpm exec projen test",
        "cwd": "packages/type-safe-api"
      }
    },
    "package": {
      "executor": "nx:run-commands",
      "options": {
        "command": "pnpm exec projen package",
        "cwd": "packages/type-safe-api"
      }
    },
    "test:watch": {
      "executor": "nx:run-commands",
      "options": {
        "command": "pnpm exec projen test:watch",
        "cwd": "packages/type-safe-api"
      }
    },
    "watch": {
      "executor": "nx:run-commands",
      "options": {
        "command": "pnpm exec projen watch",
        "cwd": "packages/type-safe-api"
      }
    },
    "eslint": {
      "executor": "nx:run-commands",
      "options": {
        "command": "pnpm exec projen eslint",
        "cwd": "packages/type-safe-api"
      }
    },
    "compat": {
      "executor": "nx:run-commands",
      "options": {
        "command": "pnpm exec projen compat",
        "cwd": "packages/type-safe-api"
      }
    },
    "package-all": {
      "executor": "nx:run-commands",
      "options": {
        "command": "pnpm exec projen package-all",
        "cwd": "packages/type-safe-api"
      }
    },
    "package:java": {
      "executor": "nx:run-commands",
      "options": {
        "command": "pnpm exec projen package:java",
        "cwd": "packages/type-safe-api"
      }
    },
    "package:python": {
      "executor": "nx:run-commands",
      "options": {
        "command": "pnpm exec projen package:python",
        "cwd": "packages/type-safe-api"
      }
    },
    "eslint-staged": {
      "executor": "nx:run-commands",
      "options": {
        "command": "pnpm exec projen eslint-staged",
        "cwd": "packages/type-safe-api"
      }
    },
    "jest": {
      "executor": "nx:run-commands",
      "options": {
        "command": "pnpm exec projen jest",
        "cwd": "packages/type-safe-api"
      }
    },
    "docgen": {
      "executor": "nx:run-commands",
      "options": {
        "command": "pnpm exec projen docgen",
        "cwd": "packages/type-safe-api"
      }
    }
  },
  "implicitDependencies": [
    "smithy-async-transformer"
  ],
  "//": "~~ Generated by projen. To modify, edit .projenrc.js and run \"pnpm exec projen\"."
}
