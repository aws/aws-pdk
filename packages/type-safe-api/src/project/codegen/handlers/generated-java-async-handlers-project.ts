/*! Copyright [Amazon.com](http://amazon.com/), Inc. or its affiliates. All Rights Reserved.
SPDX-License-Identifier: Apache-2.0 */
import {
  GeneratedJavaHandlersBaseProject,
  GeneratedJavaHandlersBaseProjectOptions,
} from "./generated-java-handlers-base-project";
import { CodegenOptions, OtherGenerators } from "../components/utils";

export interface GeneratedJavaAsyncHandlersProjectOptions
  extends GeneratedJavaHandlersBaseProjectOptions {}

export class GeneratedJavaAsyncHandlersProject extends GeneratedJavaHandlersBaseProject {
  constructor(options: GeneratedJavaAsyncHandlersProjectOptions) {
    super(options);
  }

  public buildCodegenOptions(): CodegenOptions {
    return {
      specPath: this.options.specPath,
      templateDirs: [OtherGenerators.JAVA_ASYNC_LAMBDA_HANDLERS],
      metadata: {
        srcDir: this.srcDir,
        tstDir: this.tstDir,
        packageName: this.packageName,
        runtimePackageName: this.options.generatedJavaTypes.packageName,
      },
    };
  }
}
