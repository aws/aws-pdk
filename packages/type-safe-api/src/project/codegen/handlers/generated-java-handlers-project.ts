/*! Copyright [Amazon.com](http://amazon.com/), Inc. or its affiliates. All Rights Reserved.
SPDX-License-Identifier: Apache-2.0 */
import {
  GeneratedJavaHandlersBaseProject,
  GeneratedJavaHandlersBaseProjectOptions,
} from "./generated-java-handlers-base-project";
import { GenerationOptions, OtherGenerators } from "../components/utils";

export interface GeneratedJavaHandlersProjectOptions
  extends GeneratedJavaHandlersBaseProjectOptions {}

export class GeneratedJavaHandlersProject extends GeneratedJavaHandlersBaseProject {
  constructor(options: GeneratedJavaHandlersProjectOptions) {
    super(options);
  }

  public buildOpenApiGeneratorOptions(): GenerationOptions {
    return {
      generator: "java",
      specPath: this.options.specPath,
      generatorDirectory: OtherGenerators.JAVA_LAMBDA_HANDLERS,
      srcDir: this.srcDir,
      tstDir: this.tstDir,
      normalizers: {
        KEEP_ONLY_FIRST_TAG_IN_OPERATION: true,
      },
      extraVendorExtensions: {
        "x-handlers-package": this.packageName,
        "x-runtime-package": this.options.generatedJavaTypes.packageName,
      },
      // Do not generate map/list types. Generator will use built in HashMap, ArrayList instead
      generateAliasAsModel: false,
    };
  }
}
