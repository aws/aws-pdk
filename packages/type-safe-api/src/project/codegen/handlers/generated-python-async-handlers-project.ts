/*! Copyright [Amazon.com](http://amazon.com/), Inc. or its affiliates. All Rights Reserved.
SPDX-License-Identifier: Apache-2.0 */
import {
  GeneratedPythonHandlersBaseProject,
  GeneratedPythonHandlersBaseProjectOptions,
} from "./generated-python-handlers-base-project";
import { GenerationOptions, OtherGenerators } from "../components/utils";

export interface GeneratedPythonAsyncHandlersProjectOptions
  extends GeneratedPythonHandlersBaseProjectOptions {}

export class GeneratedPythonAsyncHandlersProject extends GeneratedPythonHandlersBaseProject {
  constructor(options: GeneratedPythonAsyncHandlersProjectOptions) {
    super(options);
  }

  public buildOpenApiGeneratorOptions(): GenerationOptions {
    return {
      generator: "python-nextgen",
      specPath: this.options.specPath,
      generatorDirectory: OtherGenerators.PYTHON_ASYNC_LAMBDA_HANDLERS,
      // Tell the generator where python source files live
      srcDir: this.moduleName,
      tstDir: this.tstDir,
      additionalProperties: {
        packageName: this.moduleName,
        projectName: this.name,
      },
      normalizers: {
        KEEP_ONLY_FIRST_TAG_IN_OPERATION: true,
      },
      extraVendorExtensions: {
        "x-runtime-module-name": this.options.generatedPythonTypes.moduleName,
      },
    };
  }
}
