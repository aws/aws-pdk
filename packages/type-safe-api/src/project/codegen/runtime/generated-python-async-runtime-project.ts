/*! Copyright [Amazon.com](http://amazon.com/), Inc. or its affiliates. All Rights Reserved.
SPDX-License-Identifier: Apache-2.0 */
import {
  GeneratedPythonRuntimeBaseProject,
  GeneratedPythonRuntimeBaseProjectOptions,
} from "./generated-python-runtime-base-project";
import { GenerationOptions, OtherGenerators } from "../components/utils";

/**
 * Configuration for the generated python types project
 */
export interface GeneratedPythonAsyncRuntimeProjectOptions
  extends GeneratedPythonRuntimeBaseProjectOptions {}

/**
 * Python project containing types generated using OpenAPI Generator CLI
 */
export class GeneratedPythonAsyncRuntimeProject extends GeneratedPythonRuntimeBaseProject {
  constructor(options: GeneratedPythonAsyncRuntimeProjectOptions) {
    super(options);
  }

  protected buildOpenApiGeneratorOptions(): GenerationOptions {
    return {
      generator: "python",
      specPath: this.options.specPath,
      generatorDirectory: OtherGenerators.PYTHON_ASYNC_RUNTIME,
      additionalProperties: {
        packageName: this.moduleName,
        projectName: this.name,
      },
      // Tell the generator where python source files live
      srcDir: this.moduleName,
      normalizers: {
        KEEP_ONLY_FIRST_TAG_IN_OPERATION: true,
      },
    };
  }
}
