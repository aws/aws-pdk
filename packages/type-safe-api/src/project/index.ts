/*! Copyright [Amazon.com](http://amazon.com/), Inc. or its affiliates. All Rights Reserved.
SPDX-License-Identifier: Apache-2.0 */
export * from "./model/type-safe-api-model-build";
export * from "./model/type-safe-api-async-model-build";
export * from "./model/smithy/smithy-project-definition";
export * from "./model/openapi/open-api-project-definition";
export * from "./type-safe-api-project";
export * from "./model/type-safe-api-model-project";
export * from "./model/smithy/smithy-definition";
export * from "./model/openapi/open-api-definition";
export * from "./type-safe-websocket-api-project";
export * from "./model/type-safe-websocket-api-model-project";
export * from "./model/smithy/smithy-async-definition";
export * from "./model/openapi/open-api-async-definition";
export * from "./model/smithy/types";
export * from "./model/openapi/open-api-model-project";
export * from "./model/openapi/open-api-async-model-project";
export * from "./model/smithy/smithy-async-model-project";
export * from "./model/smithy/smithy-model-project";
export * from "./languages";
export * from "./types";
export * from "./typescript-project-options";
export * from "./python-project-options";
export * from "./java-project-options";
