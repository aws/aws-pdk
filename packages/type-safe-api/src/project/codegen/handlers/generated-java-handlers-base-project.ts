/*! Copyright [Amazon.com](http://amazon.com/), Inc. or its affiliates. All Rights Reserved.
SPDX-License-Identifier: Apache-2.0 */
import * as path from "path";
import { DependencyType, SampleDir } from "projen";
import { JavaProject } from "projen/lib/java";
import { JavaVersion } from "../../languages";
import {
  CodeGenerationSourceOptions,
  GeneratedJavaHandlersOptions,
} from "../../types";
import { OpenApiGeneratorHandlebarsIgnoreFile } from "../components/open-api-generator-handlebars-ignore-file";
import { OpenApiGeneratorIgnoreFile } from "../components/open-api-generator-ignore-file";
import { OpenApiToolsJsonFile } from "../components/open-api-tools-json-file";
import { TypeSafeApiCommandEnvironment } from "../components/type-safe-api-command-environment";
import {
  buildCleanOpenApiGeneratedCodeCommand,
  buildInvokeOpenApiGeneratorCommandArgs,
  buildTypeSafeApiExecCommand,
  GenerationOptions,
  TypeSafeApiScript,
} from "../components/utils";
import { GeneratedJavaRuntimeBaseProject } from "../runtime/generated-java-runtime-base-project";
import { RuntimeVersionUtils } from "../runtime-version-utils";

export interface GeneratedJavaHandlersBaseProjectOptions
  extends GeneratedJavaHandlersOptions,
    CodeGenerationSourceOptions {
  /**
   * The generated java types
   */
  readonly generatedJavaTypes: GeneratedJavaRuntimeBaseProject;
}

export abstract class GeneratedJavaHandlersBaseProject extends JavaProject {
  /**
   * Options configured for the project
   */
  protected readonly options: GeneratedJavaHandlersBaseProjectOptions;

  /**
   * Source directory
   */
  protected readonly srcDir: string;

  /**
   * Test directory
   */
  protected readonly tstDir: string;

  /**
   * Java package name
   */
  public readonly packageName: string;

  /**
   * Java runtime version for the handlers
   */
  public readonly runtimeVersion: JavaVersion;

  constructor(options: GeneratedJavaHandlersBaseProjectOptions) {
    super({
      sample: false,
      junit: false,
      compileOptions: RuntimeVersionUtils.JAVA.getMavenCompileOptions(
        options.runtimeVersion
      ),
      ...(options as any),
    });
    TypeSafeApiCommandEnvironment.ensure(this);
    this.options = options;
    this.runtimeVersion = options.runtimeVersion ?? JavaVersion.JAVA_17;
    this.packageName = `${this.pom.groupId}.${this.name}.handlers`;
    this.srcDir = path.join(
      "src",
      "main",
      "java",
      ...this.packageName.split(".")
    );
    this.tstDir = path.join(
      "src",
      "test",
      "java",
      ...this.packageName.split(".")
    );

    [
      `${options.generatedJavaTypes.pom.groupId}/${options.generatedJavaTypes.pom.artifactId}@${options.generatedJavaTypes.pom.version}`,
    ]
      .filter(
        (dep) =>
          !this.deps.tryGetDependency(dep.split("@")[0], DependencyType.RUNTIME)
      )
      .forEach((dep) => this.addDependency(dep));

    // Pin constructs version
    this.deps.removeDependency(
      "software.constructs/constructs",
      DependencyType.BUILD
    );
    this.addDependency("software.constructs/constructs@10.3.0");

    // Dependency on junit for tests
    this.deps.addDependency(
      "org.junit.jupiter/junit-jupiter-engine@5.10.2",
      DependencyType.TEST
    );

    // Remove the projen test dependency since otherwise it takes precedence, causing projen to be unavailable at synth time
    this.deps.removeDependency("io.github.cdklabs/projen", DependencyType.TEST);

    // Add a dependency on the generated java types repository
    this.pom.addRepository({
      url: `file://${path.relative(
        this.outdir,
        options.generatedJavaTypes.outdir
      )}/dist/java`,
      id: `${options.generatedJavaTypes.pom.groupId}-${options.generatedJavaTypes.pom.artifactId}-repo`,
    });

    // Ignore everything for the first mustache pass
    const openapiGeneratorIgnore = new OpenApiGeneratorIgnoreFile(this);
    openapiGeneratorIgnore.addPatterns("/*", "**/*", "*");
    // Ignore everything but the handler files for the handlebars pass
    const openapiGeneratorHandlebarsIgnore =
      new OpenApiGeneratorHandlebarsIgnoreFile(this);
    openapiGeneratorHandlebarsIgnore.addPatterns(
      "/*",
      "**/*",
      "*",
      // This will be split into a file per targeted handler
      `!${this.srcDir}/__all_handlers.java`,
      `!${this.tstDir}/__all_tests.java`
    );

    // Add OpenAPI Generator cli configuration
    OpenApiToolsJsonFile.ensure(this).addOpenApiGeneratorCliConfig(
      options.openApiGeneratorCliConfig
    );

    const generateTask = this.addTask("generate");
    generateTask.exec(buildCleanOpenApiGeneratedCodeCommand());
    generateTask.exec(
      buildTypeSafeApiExecCommand(
        TypeSafeApiScript.GENERATE,
        this.buildGenerateCommandArgs()
      )
    );

    this.preCompileTask.spawn(generateTask);

    // Ignore the openapi generator metadata files
    this.gitignore.addPatterns(".openapi-generator");

    // Use the maven shade plugin to build a "super jar" which we can deploy to AWS Lambda
    this.pom.addPlugin("org.apache.maven.plugins/maven-shade-plugin@3.3.0", {
      configuration: {
        createDependencyReducedPom: false,
        transformers: [
          {
            // Transformer required for merging log4j2 plugins cache file
            // https://docs.aws.amazon.com/lambda/latest/dg/java-logging.html#java-logging-cdk
            transformer: {
              "@implementation":
                "com.github.edwgiz.maven_shade_plugin.log4j2_cache_transformer.PluginsCacheFileTransformer",
            },
          },
        ],
      },
      executions: [
        {
          id: "shade-task",
          phase: "package",
          goals: ["shade"],
        },
      ],
      dependencies: [
        "com.github.edwgiz/maven-shade-plugin.log4j2-cachefile-transformer@2.15.0",
      ],
    });

    // Use the maven surefire plugin to run junit tests
    this.pom.addPlugin("org.apache.maven.plugins/maven-surefire-plugin@3.2.1");

    // Log4j2 configuration for powertools logger
    new SampleDir(this, "src/main/resources", {
      files: {
        "log4j2.xml": `<?xml version="1.0" encoding="UTF-8"?>
<Configuration>
    <Appenders>
        <Console name="JsonAppender" target="SYSTEM_OUT">
            <JsonTemplateLayout eventTemplateUri="classpath:LambdaJsonLayout.json" />
        </Console>
    </Appenders>
    <Loggers>
        <Logger name="JsonLogger" level="INFO" additivity="false">
            <AppenderRef ref="JsonAppender"/>
        </Logger>
        <Root level="info">
            <AppenderRef ref="JsonAppender"/>
        </Root>
    </Loggers>
</Configuration>`,
      },
    });
  }

  public buildGenerateCommandArgs = () => {
    return buildInvokeOpenApiGeneratorCommandArgs(
      this.buildOpenApiGeneratorOptions()
    );
  };

  public abstract buildOpenApiGeneratorOptions(): GenerationOptions;
}
