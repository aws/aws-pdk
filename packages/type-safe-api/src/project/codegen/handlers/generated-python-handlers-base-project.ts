/*! Copyright [Amazon.com](http://amazon.com/), Inc. or its affiliates. All Rights Reserved.
SPDX-License-Identifier: Apache-2.0 */
import * as path from "path";
import { DependencyType, SampleFile } from "projen";
import { PythonProject } from "projen/lib/python";
import { PythonVersion } from "../../languages";
import {
  CodeGenerationSourceOptions,
  GeneratedPythonHandlersOptions,
  Architecture,
} from "../../types";
import { OpenApiGeneratorHandlebarsIgnoreFile } from "../components/open-api-generator-handlebars-ignore-file";
import { OpenApiGeneratorIgnoreFile } from "../components/open-api-generator-ignore-file";
import { OpenApiToolsJsonFile } from "../components/open-api-tools-json-file";
import { TypeSafeApiCommandEnvironment } from "../components/type-safe-api-command-environment";
import {
  buildCleanOpenApiGeneratedCodeCommand,
  buildInvokeOpenApiGeneratorCommandArgs,
  buildTypeSafeApiExecCommand,
  GenerationOptions,
  TypeSafeApiScript,
} from "../components/utils";
import { GeneratedPythonRuntimeBaseProject } from "../runtime/generated-python-runtime-base-project";
import { RuntimeVersionUtils } from "../runtime-version-utils";

export interface GeneratedPythonHandlersBaseProjectOptions
  extends GeneratedPythonHandlersOptions,
    CodeGenerationSourceOptions {
  /**
   * The generated python types
   */
  readonly generatedPythonTypes: GeneratedPythonRuntimeBaseProject;
}

export abstract class GeneratedPythonHandlersBaseProject extends PythonProject {
  /**
   * Options configured for the project
   */
  protected readonly options: GeneratedPythonHandlersBaseProjectOptions;

  /**
   * Directory containing tests
   */
  protected readonly tstDir: string;

  /**
   * Python runtime version for the handlers
   */
  public readonly runtimeVersion: PythonVersion;

  constructor(options: GeneratedPythonHandlersBaseProjectOptions) {
    super({
      pytest: true,
      poetryOptions: {
        packages: [{ include: options.moduleName }],
        // Module must be explicitly added to include since poetry excludes everything in .gitignore by default
        include: [options.moduleName!, `${options.moduleName}/**/*.py`],
      },
      ...(options as any),
      sample: false,
      poetry: true,
    });
    TypeSafeApiCommandEnvironment.ensure(this);
    this.options = options;
    this.runtimeVersion = options.runtimeVersion ?? PythonVersion.PYTHON_3_11;
    this.tstDir = "test";

    if (options.pytest ?? true) {
      // Pytest fails with exit code 5 when there are no tests.
      // We want to allow users to delete all their tests, or to upgrade an existing project without breaking their build
      // See: https://github.com/pytest-dev/pytest/issues/2393
      this.testTask.reset(
        `pytest ${this.tstDir}/ || ([ $? = 5 ] && exit 0 || exit $?)`
      );
    }

    [
      RuntimeVersionUtils.PYTHON.getPythonDependencyVersion(
        this.runtimeVersion
      ),
      `${options.generatedPythonTypes.name}@{path="${path.relative(
        this.outdir,
        options.generatedPythonTypes.outdir
      )}", develop=true}`,
    ]
      .filter((dep) => !this.deps.tryGetDependency(dep, DependencyType.RUNTIME))
      .forEach((dep) => this.addDependency(dep));

    // Ignore everything for the first mustache pass
    const openapiGeneratorIgnore = new OpenApiGeneratorIgnoreFile(this);
    openapiGeneratorIgnore.addPatterns("/*", "**/*", "*");
    // Ignore everything but the handler files for the handlebars pass
    const openapiGeneratorHandlebarsIgnore =
      new OpenApiGeneratorHandlebarsIgnoreFile(this);
    openapiGeneratorHandlebarsIgnore.addPatterns(
      "/*",
      "**/*",
      "*",
      // This will be split into a file per targeted handler
      `!${this.moduleName}/__all_handlers.py`,
      `!${this.tstDir}/__all_tests.py`
    );

    // Add OpenAPI Generator cli configuration
    OpenApiToolsJsonFile.ensure(this).addOpenApiGeneratorCliConfig({
      version: "6.6.0",
      ...options.openApiGeneratorCliConfig,
    });

    const generateTask = this.addTask("generate");
    generateTask.exec(buildCleanOpenApiGeneratedCodeCommand());
    generateTask.exec(
      buildTypeSafeApiExecCommand(
        TypeSafeApiScript.GENERATE,
        this.buildGenerateCommandArgs()
      )
    );

    this.preCompileTask.spawn(generateTask);

    // Ignore the generated code
    this.gitignore.addPatterns(".openapi-generator");

    // Write __init__.py as sample code
    new SampleFile(this, path.join(this.moduleName, "__init__.py"), {
      contents: "#",
    });

    // Package task to build a distributable which can be deployed to lambda
    this.packageTask.exec("mkdir -p dist/lambda && rm -rf dist/lambda/*");
    this.packageTask.exec(
      `cp -r ${this.moduleName} dist/lambda/${this.moduleName}`
    );
    this.packageTask.exec(
      "poetry export --without-hashes --format=requirements.txt | sed -E 's/^-e[[:space:]]+//' > dist/lambda/requirements.txt"
    );
    // Select the platform based on the specified architecture, defaulting to x86_64
    // See: https://docs.aws.amazon.com/lambda/latest/dg/python-package.html#python-package-native-libraries
    const platform =
      options.architecture === Architecture.ARM_64
        ? "manylinux2014_aarch64"
        : "manylinux2014_x86_64";
    this.packageTask.exec(
      `pip install -r dist/lambda/requirements.txt --target dist/lambda --upgrade --platform ${platform} --only-binary :all: --python-version ${RuntimeVersionUtils.PYTHON.getPipPackagingPythonVersion(
        this.runtimeVersion
      )}`
    );
  }

  public buildGenerateCommandArgs = () => {
    return buildInvokeOpenApiGeneratorCommandArgs(
      this.buildOpenApiGeneratorOptions()
    );
  };

  public abstract buildOpenApiGeneratorOptions(): GenerationOptions;
}
