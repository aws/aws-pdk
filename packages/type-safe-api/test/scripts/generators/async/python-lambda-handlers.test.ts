/*! Copyright [Amazon.com](http://amazon.com/), Inc. or its affiliates. All Rights Reserved.
SPDX-License-Identifier: Apache-2.0 */
import os from "os";
import * as path from "path";
import { exec } from "projen/lib/util";
import { GeneratedPythonAsyncHandlersProject } from "../../../../src/project/codegen/handlers/generated-python-async-handlers-project";
import { GeneratedPythonAsyncRuntimeProject } from "../../../../src/project/codegen/runtime/generated-python-async-runtime-project";
import { withTmpDirSnapshot } from "../../../project/snapshot-utils";

describe("Python Async Handlers Code Generation Script Unit Tests", () => {
  it.each(["handlers.yaml"])("Generates With %s", (spec) => {
    const specPath = path.resolve(
      __dirname,
      `../../../resources/specs/async/${spec}`
    );

    const snapshot = withTmpDirSnapshot(os.tmpdir(), (outdir) => {
      exec(`cp ${specPath} ${outdir}/spec.yaml`, {
        cwd: path.resolve(__dirname),
      });
      const clientOutdir = path.join(outdir, "client");
      const client = new GeneratedPythonAsyncRuntimeProject({
        name: "test-client",
        moduleName: "test_client",
        authorEmail: "me@example.com",
        authorName: "test",
        version: "1.0.0",
        outdir: clientOutdir,
        specPath: "../spec.yaml",
      });
      const handlersOutdir = path.join(outdir, "handlers");
      const project = new GeneratedPythonAsyncHandlersProject({
        name: "test-handlers",
        moduleName: "test_handlers",
        authorEmail: "me@example.com",
        authorName: "test",
        version: "1.0.0",
        outdir: handlersOutdir,
        specPath: "../spec.yaml",
        generatedPythonTypes: client,
      });
      project.synth();
      exec(
        `${path.resolve(
          __dirname,
          "../../../../scripts/type-safe-api/run.js generate"
        )} ${project.buildGenerateCommandArgs()}`,
        {
          cwd: handlersOutdir,
        }
      );
    });

    expect(snapshot["handlers/test_handlers/python_one.py"]).toMatchSnapshot();
    expect(snapshot["handlers/test_handlers/python_two.py"]).toMatchSnapshot();
    expect(snapshot["handlers/test/test_python_one.py"]).toMatchSnapshot();
    expect(snapshot["handlers/test/test_python_two.py"]).toMatchSnapshot();

    // Other language handlers should be skipped
    expect(
      Object.keys(snapshot).filter((p) =>
        p.toLowerCase().includes("typescript")
      )
    ).toHaveLength(0);
    expect(
      Object.keys(snapshot).filter((p) => p.toLowerCase().includes("java"))
    ).toHaveLength(0);

    // Split file should be deleted
    expect(
      Object.keys(snapshot).filter((p) =>
        p.toLowerCase().includes("__all_handlers")
      )
    ).toHaveLength(0);
  });
});
