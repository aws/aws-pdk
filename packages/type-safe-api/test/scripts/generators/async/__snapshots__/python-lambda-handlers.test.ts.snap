// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Python Async Handlers Code Generation Script Unit Tests Generates With handlers.yaml 1`] = `
"from test_client.models import *
from test_client.response import Response
from test_client.interceptors import INTERCEPTORS
from test_client.interceptors.powertools.logger import LoggingInterceptor
from test_client.api.operation_config import (
    python_one_handler, PythonOneRequest, PythonOneOperationResponses
)


def python_one(input: PythonOneRequest, **kwargs) -> PythonOneOperationResponses:
    """
    Type-safe handler for the PythonOne operation
    """
    LoggingInterceptor.get_logger(input).info("Start PythonOne Operation")

    # TODO: Implement PythonOne Operation. \`input\` contains the request input

    return Response.internal_failure(InternalFailureErrorResponseContent(
        message="Not Implemented!"
    ))


# Entry point for the AWS Lambda handler for the PythonOne operation.
# The python_one_handler method wraps the type-safe handler and manages marshalling inputs and outputs
handler = python_one_handler(interceptors=INTERCEPTORS)(python_one)

"
`;

exports[`Python Async Handlers Code Generation Script Unit Tests Generates With handlers.yaml 2`] = `
"from test_client.models import *
from test_client.response import Response
from test_client.interceptors import INTERCEPTORS
from test_client.interceptors.powertools.logger import LoggingInterceptor
from test_client.api.operation_config import (
    python_two_handler, PythonTwoRequest, PythonTwoOperationResponses
)


def python_two(input: PythonTwoRequest, **kwargs) -> PythonTwoOperationResponses:
    """
    Type-safe handler for the PythonTwo operation
    """
    LoggingInterceptor.get_logger(input).info("Start PythonTwo Operation")

    # TODO: Implement PythonTwo Operation. \`input\` contains the request input

    return Response.internal_failure(InternalFailureErrorResponseContent(
        message="Not Implemented!"
    ))


# Entry point for the AWS Lambda handler for the PythonTwo operation.
# The python_two_handler method wraps the type-safe handler and manages marshalling inputs and outputs
handler = python_two_handler(interceptors=INTERCEPTORS)(python_two)

"
`;

exports[`Python Async Handlers Code Generation Script Unit Tests Generates With handlers.yaml 3`] = `
"import pytest
from aws_lambda_powertools import Logger

from test_handlers.python_one import python_one
from test_client.api.operation_config import (
    PythonOneRequest, PythonOneRequestParameters, PythonOneRequestBody
)


@pytest.fixture
def request_arguments():
    """
    Fixture for constructing common request arguments
    """
    return {
        "event": {},
        "context": None,
        "interceptor_context": {
            "logger": Logger(),
        },
    }


def test_python_one_should_return_not_implemented_error(request_arguments):
    # TODO: Update the test as appropriate when you implement your handler
    response = python_one(PythonOneRequest(
        **request_arguments,
        # request_parameters=PythonOneRequestParameters(
        #     # Add request parameters here...
        # ),
        request_parameters=None,
        # body=PythonOneRequestBody(
        #     # Add body fields here...
        # )
        body=None,
    ))

    assert response.status_code == 500
    assert response.body.message == "Not Implemented!"

"
`;

exports[`Python Async Handlers Code Generation Script Unit Tests Generates With handlers.yaml 4`] = `
"import pytest
from aws_lambda_powertools import Logger

from test_handlers.python_two import python_two
from test_client.api.operation_config import (
    PythonTwoRequest, PythonTwoRequestParameters, PythonTwoRequestBody
)


@pytest.fixture
def request_arguments():
    """
    Fixture for constructing common request arguments
    """
    return {
        "event": {},
        "context": None,
        "interceptor_context": {
            "logger": Logger(),
        },
    }


def test_python_two_should_return_not_implemented_error(request_arguments):
    # TODO: Update the test as appropriate when you implement your handler
    response = python_two(PythonTwoRequest(
        **request_arguments,
        # request_parameters=PythonTwoRequestParameters(
        #     # Add request parameters here...
        # ),
        request_parameters=None,
        # body=PythonTwoRequestBody(
        #     # Add body fields here...
        # )
        body=None,
    ))

    assert response.status_code == 500
    assert response.body.message == "Not Implemented!"

"
`;
